### DISCUSSION TITLE
Memory-friendly per-query disk caching API

### DISCUSSION THREAD

**Discussion by scarlac** [TanStack/query#2649] [Ideas] [answered]
The current persistor API and experimental implementations for both web and React Native will work for minor use cases, however the general approach isn't easy to make performant as it assumes the entire state should be persisted on every change and is always available after initial restoration.

The current API looks like this:
```typescript
export interface Persistor {
    persistClient(persistClient: PersistedClient): Promisable<void>;
    restoreClient(): Promisable<PersistedClient | undefined>;
    removeClient(): Promisable<void>;
}
```

Even when throttling the `persistClient()` function, the entire app state can easily be several megabytes of data. If any part of your app changes this causes several megabytes worth of JSON serialization which is unfortunately very costly in both terms of CPU, battery use, and as a result the UX will degrade.

The issue isn't unique to React Query though. Redux Persist tries to mitigate this issue by allowing setting a custom persistor for each reducer. This approach is also not sustainable when 1 payload can be 1 MB, and you potentially need to store several.

On web (let's face it: Desktop machines) apps typically are short lived, counted in seconds and perhaps a minute or two. In React Native, apps can live on for hours, and with a persistor, days. In these situations it's favorable that we have a better API for persisting according to the cache key of the data.

E.g. (pseudo-code) instead of caching as such:
```typescript
await AsyncStorage.setItem('REACT_QUERY_OFFLINE_CACHE', JSON.stringify(client));
```
You'd store each piece of data as such:
```typescript
for(const query of clientState.queries) {
  await AsyncStorage.setItem('RC_${query.queryHash}', JSON.stringify(query.data));
}
```

This _can_ be implemented using the current API, but it would cause re-writing of the entire cache, for every query data entry. It makes the app more responsive because every item is awaited but overall performance is roughly the same (or slightly worse because of await having a slight overhead).

This general challenge is not unique to React Query. Redux Persist has it too. They allow breaking reducers into separate storage configurations such that you can store data separately and avoid having to re-serialize everything even for small changes. However, this approach also doesn't address storage of individual items - you have to persist the entire reducer so if your data is stored as such:
```
myReducer: {
  byId: {
    'id-42': { id: 42, username: 'scarlac' },
    'id-21': { id: 21, username: 'example' },
    // ... [and 1000 more items]
  }
}
```
Then you need to re-serialize myReducer in it's entirety even if only `id-42` changes.
Ideally this would also allow the cache to be filled on-demand, such that `byId` could start out empty (`byId: {}`) and then get filled in by calling the persistor API to check if there is a disk cache item with that particular key before triggering a refetch. 

I understand that the current API is experimental. I think now is a fantastic time to consider this use case for mobile apps where memory is severely constrained but disk usage isn't.

**Comment by henrymoulton** [comment]
@scarlac I was also thinking along these lines and noticed https://github.com/tannerlinsley/react-query/pull/2864 was merged to potentially allow for custom functions for serialisation/deserialisation. Would that help?

**Reply by scarlac** [reply]
@henrymoulton Unfortunately it wouldn't work well because the serialize/deserialize is a store-everything approach. With a big cache this becomes a performance issue/it doesn't scale. The real underlying problem is that the queryCache is synchronous and assumed to always be in memory. If you really truly use RQ on a big application you'll find that you end up storing too much in memory and that becomes the problem: There is no way to "off-load" to a permanent storage, let alone have the permanent storage be garbage collected (because the cacheTime GC assumes it's in memory as well).

I actually discussed this with Tanner and he suggested to hook into the query events, which allows me to selectively create, inject, and store to disk. However, it breaks anything relating to cacheTime, and there are other downsides as well (e.g. you have to manually delete query data while still keeping metadata, otherwise some queries may break).

**Comment by gabrielbull** [comment]
I'm having the same issue as well. I cannot use react-query's persistor on a native app because it will use too much memory and the OS will terminate the app. Since most native apps need to be offline-first, this makes react-query unusable in my case. It's unfortunate as I really like the library but for now I cannot use it and directly use axios with realm.js.

**Reply by henrymoulton** [reply]
@gabrielbull how much memory is this app that you're using?

**Reply by gabrielbull** [reply]
The data itself is not that big, sometimes only a few MBs. It seems to get exponential when used with react-query though, a few MBs of data = hundreds of MBs of memory usage.

**Comment by TkDodo** [comment]
@scarlac I'm open for suggestions on how to improve this for v4. Just  so that I understand what we'd want here:

- still store the whole cache / whichever queries are chosen via `dehydrateOptions`, but store them on a per-query basis
- we'd still want to restore all entries on page load? Note that I have some changes for that api here: https://github.com/tannerlinsley/react-query/pull/3248
- when one query updates, we don't write everything, but only the affected entries

I guess this wouldn't alleviate the memory pain much, as we would still have the cache in memory. I guess unused queries could be cleared out by setting `cacheTime` - but I assume this would also delete it from the persisted cache?

Maybe we would need to find a way to keep queries persisted while having them cleared from memory. I'm open to suggestions here :)

---

@gabrielbull react-query only keeps one copy per query of your data in memory, so I'm not sure how a few MB of data can lead to hundreds of MB memory usage 🤔

**Reply by henrymoulton** [reply]
@TkDodo I think the core of @scarlac's suggestion "however the general approach isn't easy to make performant as it assumes the entire state should be persisted on every change and is always available after initial restoration."

is the concept of adding "Lazy loading" of queries from the persistor. Case-studies to perhaps look at are [WatermelonDB](https://github.com/Nozbe/WatermelonDB) or [Realm](https://docs.mongodb.com/realm/sdk/react-native/):

> For simple apps, using Redux or MobX with a persistence adapter is the easiest way to go. But when you start scaling to thousands or tens of thousands of database records, your app will now be slow to launch (especially on slower Android devices). Loading a full database into JavaScript is expensive! Watermelon fixes it by being lazy. Nothing is loaded until it's requested. And since all querying is performed directly on the rock-solid SQLite database on a separate native thread, most queries resolve in an instant.
https://github.com/Nozbe/WatermelonDB

> Realm Database only runs a query when you request the results of that query. This lazy evaluation enables you to write elegant, highly-performant code for handling large data sets and complex queries.
https://docs.mongodb.com/realm/sdk/react-native/data-types/collections/


Thinking out loud an API might look like

```
// setup
 persistQueryClient({
   queryClient,
   persistor: localStoragePersistor,
   lazy: true
 })

// later 
const queryName = "get-foos"

persistQueryClient.loadFromPersistence(queryName)
```

Thanks for the response @TkDodo and keen to hear what you think!

**Reply by TkDodo** [reply]
One problem with lazily reading from an async storage is that you will get hard loading states while you are trying to read - which is what restoration from that stores tries to solve. If you want that, why not read from async storage in the queryFn if the current data in the cache is undefined, and only go to the server if there is also nothing in that storage

**Reply by scarlac** [reply]
True, @TkDodo 

The loading states for any query, also cached queries, is expected behavior when you cacher is async. In practice it's however so brief that it rarely becomes visible to the user. In my use case (a mobile app) the navigation screen transitions cover it up and the UI is as responsive as if it was synchronous.

The query persistor is really just 1 piece of the puzzle. There are a few separate pieces:
1. Write query to disk cache (persisting)
2. Purge inactive queries from memory (memory use reduction)
3. Restore query from cache ONLY when needed (current persistor approach assumes everything at startup)

Lazy loading as a concept is fine but normally you wouldn't have to worry about purging memory - I am proposing that memory can be freed so long running apps can remain stable and fast over time.

**Reply by TkDodo** [reply]
I'm trying to wrap my head around how this should work in RQ, but I haven't found anything really. One approach would be to do this on the `QueryCache` level, maybe provide an `AsyncQueryCache` that could have a limited memory store and always store everything on an async store. However, it would make reading from the cache per se async, as a value could _not_ be in memory and must therefore be retrieved from the AsyncStorage before. We currently rely on multiple places on the fact that we can read synchronously from the cache, so I don't think that's an option.

With v4, you can subscribe / write to / read from with the persistPlugin separately. as we have split up the functions: https://react-query-alpha.tanstack.com/plugins/persistQueryClient. This would help in lazy-loading into the cache at a later time than the startup. 

With that, I think a query-level persistence rather than a full query-client level persistence should be possible by just spawning multiple subscriptions, each one only syncing a single query to a separate storage name. However, restoration would need to be done before you run the query for the first time, so I'm not sure where we could "hook into" right now.

But all of these don't solve the problem of purging memory. Do you have a tangible proposal for how that could look like?

**Reply by scarlac** [reply]
A tangible proposal, no. I simply don't have enough experience.

One thing I kept coming back to was that observers must have queries available which blocks them from being async. It's fair that actively observed queries are in memory, but currently inactive queries must also exist. That's mainly where the wheels come off. If we can find a way to restore queries on demand, when an query returns, then I think we can get by with observing the store manually and hard deleting unused (0 observers) queries.

Once we have polished the workaround, I can share it so you can see how it falls short, if this explainer wasn't clear.

**Reply by scarlac** [reply]
Here's the code:
https://gist.github.com/scarlac/34f5bd183d849086654400367817f09b

Note that this doesn't actually work entirely as intended, as unused queries metadata is still kept in memory. As is apparent from the code, you can also see that `cacheTime` no longer serves a purpose because we never want queries to linger around in memory - they should be purged.

**Reply by henrymoulton** [reply]
> However, it would make reading from the cache per se async, as a value could not be in memory and must therefore be retrieved from the AsyncStorage before. We currently rely on multiple places on the fact that we can read synchronously from the cache, so I don't think that's an option.

@TkDodo I'm currently using [react-native-mmkv](https://github.com/mrousavy/react-native-mmkv) which is synchronous. Right now I actually have to turn the calls async because `storage` assumes an async persistor. I think this is quite a common trend in React Native land which is to make more libraries synchronous. The new architecture for React Native is generally more synchronous e.g. Fabric https://reactnative.dev/docs/fabric-renderer

```
const MmkvStorage = {
  getItem: (key: string) => {
    const result = mmkvStore.getString(key); // Note(Henry): no await here, this is synchronous!
    return Promise.resolve(result ? result : null);
  },
  setItem: (key: string, value: string) => {
    mmkvStore.set(key, value);
    return Promise.resolve();
  },
  removeItem: (key: string) => {
    mmkvStore.delete(key);
    return Promise.resolve();
  },
};

export const asyncStoragePersistor = createAsyncStoragePersistor({
  storage: MmkvStorage,
});
```

What do you think about this?

**Reply by TkDodo** [reply]
interesting. I think `perstistQueryClient` is async because it is easy to make something that is synchronous async, but not the other way around. localStorage is actually also sync, but things like [localForage](https://github.com/localForage/localForage) is not because it can also read from IndexDB.

**Reply by TkDodo** [reply]
@scarlac I'm afraid this won't work on a bunch of levels:
1) you're mutating the query internal state, which is not advised.
2) `undefined` will very likely be a disallowed state value in v4 for successful queries
3) I'm not sure how this works with `cacheTime: 0`: Once you get the `observerRemoved` event, the `cacheTime: 0` should have already removed the query completely, so why do you need to set the data to `undefined` - and why does the meta data stay behind in the cache as you say?

**Reply by scarlac** [reply]
There isn't a trend to move to synchronous code in RN. MMKV uses an unofficial API (JSI) and also uses an in memory cache as well, as well as writing to disk. I would not advise a  sync persistence API.

In regards to the example code, that actually works right, in production, with the mentioned caveats. RQ uses undefined itself already, so I'm not sure how it could be removed. If a query doesn't exist, what does RQ v4 then return on the first render?

Reg cacheTime:0, yes this is because of the way cacheTime works: it's not the total cache time, it's relative to whenever a query becomes unused. I had to read the source to fully understand this but if set to 0, the cache remains in memory until all observers are gone, and all outstanding promises are resolved. Once resolved, it's garbage collected using a simply setTimeout and cacheTime. At least, this is how I understand it.
The metadata doesn't stay behind by default: I have to leave it behind, otherwise outstanding promises with correct results will not be used. If a promise resolves from the server and the query is entirely gone, it doesn't have the context to recreate the query or save the value to disk, therefore the metadata must be kept to ensure that slow queries are cached, even if there are no observers when they return.

**Reply by TkDodo** [reply]
> RQ uses undefined itself already, so I'm not sure how it could be removed. If a query doesn't exist, what does RQ v4 then return on the first render?

I specifically said for _successful_ queries - so queries that are in the _success_state. Of course loading queries will return _undefined_. You can read here for the reasoning behind it: https://github.com/tannerlinsley/react-query/issues/2925

> Reg cacheTime:0, yes this is because of the way cacheTime works: it's not the total cache time, it's relative to whenever a query becomes unused. I had to read the source to fully understand this but if set to 0, the cache remains in memory until all observers are gone, and all outstanding promises are resolved. Once resolved, it's garbage collected using a simply setTimeout and cacheTime. At least, this is how I understand it.

Your understanding is correct, which is precisely why I don't understand how this works, especially this part about the `observerRemoved` event: https://gist.github.com/scarlac/34f5bd183d849086654400367817f09b#file-memory-efficient-rq-client-tsx-L106-L111

If you remove the last observer, `cacheTime: 0` will completely remove the query for you from the cache.

---

I generally like this idea, just have to get my head around how it exactly works and how we could make it a bit "less hacky" :) But it has potential to become a separate plugin, like a `LazyPersisterPlugin` or so :)

**Reply by scarlac** [reply]
Sorry, my comment on `cacheTime:0` was made mostly as a theory. In practice I have supplied random values so this has mostly been the default. I have not thoroughly tested the 0-value. It's entirely possible it doesn't work reliably. If I recall correctly, the cacheTime & GC only comes into play with 0 observers and a non-loading query. So if a query was in a loading state, the GC is postponed and re-tried later.

Yeah - the idea was actually to abstract the "RNFS" (React Native File System) stuff and make a "plugin" out of this, allowing any type of per-query persistence and memory efficient use of RQ.
Perhaps all we do here is end up agreeing that "empty queries" are okay, and then figure out how we can deal with these lingering "empty" & slow queries so we get full efficiency.

I'm generally OKAY with not making RQ's cache async if we get more control over garbage collection and somehow figure out how to deal with removing queries that are technically still in flight (e.g. so they get re-created temporarily once resolved, or any method that allows the plugin to deal with it). That's really the limitation that this "plugin" can't deal with as it stands.

**Reply by henrymoulton** [reply]
>There isn't a trend to move to synchronous code in RN. MMKV uses an unofficial API (JSI) and also uses an in memory cache as well, as well as writing to disk. I would not advise a sync persistence API.

Trying to stay on-topic but my opinion is that there is a move towards more synchronous code for persistence in RN:
https://github.com/greentriangle/react-native-leveldb
https://medium.com/@Sendbird/extreme-optimization-of-asyncstorage-in-react-native-b2a1e0107b34
https://github.com/ospfranco/react-native-quick-sqlite

and in Node too:
https://github.com/JoshuaWise/better-sqlite3

Not sure what you mean by JSI being "unofficial", but overall sounds like you might have some interesting thoughts and would love to hear them.

**Reply by scarlac** [reply]
@henrymoulton Would love to talk about it but this isn't the forum. To quote Tk for a second, it's easy to go sync -> async, but not the other way around. However right now the main issue (for this discussion) is that the arch just doesn't lend itself to be persisted at all.

@TkDodo Do you think there is a chance v4 would see changes that allowed us deeper integration?

**Reply by KannarFr** [reply]
@TkDodo here is a little reminder.

Also, if it won't be implemented in v4. I think we should warn users that `offlineFirst` mode can be very tricky without enough place to store data on mobile storage.

We should choose/define the way to implement it. I'll be glad to work on it if needed. TBH, I discovered RQv4 and rewrote all my application using `offlineFirst` and it's now "stuck" because I can't store enough required data on mobile storage (yet!).

Maybe a quick win could be storing `queryData` (only) on the mobile's filesystem when `offlineFirst` is enabled, or when a threshold is reached. Just dirty suggestions which can drive us to nicer ideas.

**Comment by jonathanstanley** [comment]
To try and narrow towards a proposal:

## Memory management
To actually reduce memory, RQ would probably need to introduce additional timers in `useQuery`.

### `memoryTime` 
If this expired, it would trigger the item to be dropped from memory.  It could remain in the persisted state, so long as `cacheTime` has not expired. However, this is tricky:

1.  Somehow the Persister needs to be able to keep the item
2. There has to be an async lookup to the persisted data... which could be tricky for state. Maybe its state must be treated just like a new query for RQ with a network call (but instead uses the local persisted cache first / offline mode).
3. Default probably should be `Infinity` for backwards compatibility

### `persistTime` 
In some cases, a user may want to keep some data forever while in an memory.  However, once persisted, the app may prefer that the persisted data isn't used after some limit (ex: 30 days).  To be clear about security: I'm not aware of any solution that can guarantee a local cache is deleted at a certain time.

## Individual persistence
By persisting each item separately instead of bundling in `PersistedClient`, we solve the `memoryTime` issue 1 above.  dropping an item from memory could be independent of whatever is kept in memory.  FWIW, [Google also advises this strategy](https://developers.google.com/web/fundamentals/instant-and-offline/web-storage/indexeddb-best-practices#keeping_your_app_performant):
> Instead of storing the entire state tree in a single record, you should break it up into individual records and only update the records that actually change.

However, this could introduce a bunch of transactions into overhead; [perhaps slowing performance in some cases](http://nparashuram.com/IndexedDB/perf/#Transaction%20based%20on%20Write).  Without knowing the use and more specific test, we won't know which is most efficient.  One way to handle would be to add an option on `useQuery` so devs can choose for themselves:

### `individualPersistence:boolean`
`false` all bundles into the main `PersistClient` (as currently done)
`true` persisted independently and not included in the main `PersistClient`

### IndexedDB

> If any part of your app changes this causes several megabytes worth of JSON serialization which is unfortunately very costly in both terms of CPU, battery use, and as a result the UX will degrade.

Persisting items individually would definitely help.  Furthermore, by using the IndexedDB persister (see #3198 ), the `stringify` penalty is avoided. It still requires a structured clone, but this should be much faster.

## Summary
In short, add three options to `useQuery`:
1. `memoryTime:number`
2. `persistTime:number`
3. `individualPersistence:boolean`

**Comment by teaguestockwell** [comment]
Instead of rehydrating queries with the persister, would it be possible to create a query cache with a persistent synchronous storage client?

What if there was a [query cache](https://github.com/tannerlinsley/react-query/blob/master/src/core/queryCache.ts) that replaced it's references to queries and queriesMap with something that: 
- could read from the disk, then rehydrate a query
- could dehydrate a query and save it to the disk

Looking at the query cache, there are a few methods that expect the queries to be hydrated:
1. remove(): calls query.destroy
1. get(): This may be able to be replaced by a lookup to disk followed by rehydrating the query?
1. getAll(): This looks like it's primarily used to show queries in the dev tools, and to determine if the query should be dehydrated. For the dev tools, we may be able to get away with only returning only the active queries as these should be small enough to fit into memory. And using the query cache as the persistence layer we would not need to persist any queries through the persister.

I'm not certain how having to rehydrate queries would effect onFocus and onOnline, or if there is a way to only have to call these methods on only active queries?

**Comment by scarlac** [comment]
@TkDodo Any major changes in v4 that could potentially carve a path for a lazy loaded query cache, and/or persisting/restoring data per-query?
I read the 'migrating to v4' article and your blog post but I don't see the offline mode as a step towards this but I did see some changes to persistence/persistor that I am hoping makes some of these things easier?

**Reply by TkDodo** [reply]
no, we didn't do this in v4 because I don't know how the apis should be yet. The persisters are designed to just sync the current QueryCache somewhere. The QueryCache currently contains all the queries in-memory.

My gut feeling says to decouple the requirements here from persisters. The persisters should stay as they are (write the whole cache). However, we can think about implementing a _different_ cache. Like, one that doesn't store everything in-memory, but only keeps a bit in-memory and writes the rest somewhere else.

Then you could do:

```diff
const queryClient = new QueryClient({
-  queryCache: new QueryCache()
+  queryCache: new AsyncQueryCache({ some options })
})
```

that cache could e.g. only keep all keys in memory, and maybe a bunch of values with a max-size; if data is not there, it would read it from an external source; the drawback would be that the query would be in hard `loading` state while data would be read, but I think this is a wanted trade-off here. What do you think?

**Reply by scarlac** [reply]
Generally speaking that sounds like the right approach @TkDodo 
The only thing I could fault, but I might be misunderstanding you:
> [...] write the whole cache [...]

I think we need a per-item cache write functionality. This could _in theory_ be implemented by the `AsyncQueryCache` class but the challenge is the before/after comparisons you'd have to do. This would cause a constant overhead where one has to iterate over all cache items, restore them from disk, compare semi-deeply (because references can't be compared), and keep doing that over and over. If each disk read+comparison takes e.g. 10ms per disk item, that quickly becomes too slow to use in practice.
So: Having AsyncQueryCache being informed of which items change would be important to achieve any kind of performance.

Anyway, I think we're on the right path then! I can't quite figure out how to structure an API that would make everyone happy here, so perhaps we can get more concrete now with your idea in mind?

**Reply by TkDodo** [reply]
> I think we need a per-item cache write functionality

Not sure about that. The idea was to keep the in-memory cache small right? like:

```
queryCache: new AsyncQueryCache({ threshold: 7 })
```

or something. Then, writing the whole cache would _only_ persist those 7 entries that are currently in memory. All other items are offloaded to the async cache storage anyways

**Comment by TkDodo** [comment]
We've created an issue to track this for v5 here:
- https://github.com/TanStack/query/issues/4693

**Comment by TkDodo** [comment, marked as answer]
We've shipped this in v5.0.0-beta.28:

https://tanstack.com/query/v5/docs/react/plugins/createPersister

PLEASE try it out and give feedback, because there's still time to change things now :)

**Reply by shashankitsoft** [reply]
@TkDodo 
It is good that v5 considered for such API. There I see the filters option to narrow down which Queries should be persisted. But the problem with this is we have to mention all query keys at one place ex App.js where we wrap <App/> root in PersistQueryClientProvider.
There can be many queries used in react app in different components where we call like `useQuery({ queryKey, queryFn, ...options })`.
**And using 'filters' to define all queryKeys at one place in top root component to include is weird. It would be better and more practical if we can pass 'persist' flag true or false in useQuery options, so that we can individually define which query data we want to be in persistent storage and which not.**
This will be useful and easier if we only want to save few query in local storage and not others (or vice-versa) specifying there itself while api/useQuery call, without need to put all those names in root component.

Is this possible?

**Reply by TkDodo** [reply]
I'm not sure what you're talking about exactly 😅 . With the new persister plugin, you can pass a persister directly to `useQuery` if you want, so it will only be for those queries. It doesn't use or need the `PersistQueryClientProvider` at all.

If you use the provider, you can use the `meta` flag to opt queries in or out of persistence

**Reply by scarlac** [reply]
@TkDodo I'm late to the after-party but I've reviewed the APIs and it's a step in the right direction! It gets us to a point where I don't have to override onFetch() but it gets us no further than that. Since we last discussed this last time, I had modified my implementation from using the internal events to using onFetch on the queryClient like so:
```ts
getQueryClient().setDefaultOptions({
      ...defaults,
      queries: {
        ...defaults.queries,
        behavior: {
          ...defaults.queries?.behavior,
          onFetch: (context) => {
            const fetchFn = context.fetchFn as () => unknown;
            context.fetchFn = async (...args) => {
            // ...code ala "await AsyncStorage.setItem(...)" goes here
```
This is less than 100 lines of code to get to the same functionality as there is in the experimental APIs, I think.

The complaints I see around `setQueryData` not persisting either seem quite legitimate. I reviewed a bunch of existing code in our code bases, and simply using the experimental API would yield the same sort of bugs as people are reporting.
To avoid rehashing that entire discussion here, I'd like to just thank you and the team for how far we've come, but just let you know that I don't think we're found "the" solution yet. I do understand the problems around making everything async, so I'll leave it at that.

**Comment by cah4a** [comment]
I have the same issue.

The more cache I want to store the more app performance drops.
Only way to have both is to have serialization and saving per query.
I throw `react-async-storage-persister` away and wrote my own implementation.

Which is working pretty flawlessly:

```ts
export const queriesStorage = new MMKV({ id: "queries" });
const queriesTimestamps = new Map<string, number>();
const xj = new XtraJson();

export const queryClientPersister: Persister = {
   persistClient: throttle((client) => {
      const time = Date.now();
      let updated = 0;
      let deleted = 0;

      queriesStorage.set("BUSTER", client.buster);
      queriesStorage.set("TIMESTAMP", client.timestamp);
      const keysToLive = new Set(["BUSTER", "TIMESTAMP"]);

      for (const query of client.clientState.queries) {
         if (query.state.status === "success" && !query.state.isInvalidated) {
            keysToLive.add(`QUERY_${query.queryHash}`);

            const updatedAt = queriesTimestamps.get(query.queryHash) ?? 0;

            if (updatedAt < query.state.dataUpdatedAt) {
               updated++;
               queriesTimestamps.set(
                  query.queryHash,
                  query.state.dataUpdatedAt,
               );
               const data = xj.stringify({
                  queryKey: query.queryKey,
                  state: query.state,
               });
               queriesStorage.set(`QUERY_${query.queryHash}`, data);
            }
         }
      }

      for (const key of queriesStorage.getAllKeys()) {
         if (!keysToLive.has(key)) {
            deleted++;
            queriesStorage.delete(key);
         }
      }

      console.log(
         `[queries] Saved in ${Date.now() - time}ms. Items count: ${keysToLive.size}. Updated: ${updated}. Deleted ${deleted}`,
      );
   }),
   restoreClient() {
      const time = Date.now();

      const result = {
         buster: queriesStorage.getString("BUSTER") ?? "",
         timestamp: queriesStorage.getNumber("TIMESTAMP") ?? 0,
         clientState: {
            queries: queriesStorage
               .getAllKeys()
               .filter((key) => key.startsWith("QUERY_"))
               .map((key) => {
                  const queryHash = key.replace("QUERY_", "");
                  const { queryKey, state } = xj.parse(
                     queriesStorage.getString(key) ?? "{}",
                  ) as {
                     queryKey: string[];
                     state: QueryState;
                  };

                  queriesTimestamps.set(queryHash, state.dataUpdatedAt);

                  return {
                     queryHash,
                     queryKey,
                     state,
                  };
               }),
            mutations: [],
         },
      };
      console.log(
         `[queries] Restored in ${Date.now() - time}ms. Items count: ${result.clientState.queries.length}`,
      );
      return result;
   },
   removeClient() {
      for (const key of queriesStorage.getAllKeys()) {
         queriesStorage.delete(key);
      }
      queriesTimestamps.clear();
      console.log("[queries] purged");
   },
};

function throttle<TArgs extends Array<unknown>>(
   func: (...args: TArgs) => void,
   wait = 100,
) {
   let timer: ReturnType<typeof setTimeout> | null = null;
   let params: TArgs;
   return function (...args: TArgs) {
      params = args;
      if (timer === null) {
         timer = setTimeout(() => {
            func(...params);
            timer = null;
         }, wait);
      }
   };
}
```

**Reply by frederikhors** [reply]
Did you tried https://tanstack.com/query/latest/docs/framework/react/plugins/createPersister?

**Reply by cah4a** [reply]
It's definitely a step forward, however it doesn't solve all issues for me.

It doesn't know about all queries and can't effectively purge stale data.
Storage size will never goes down in size as maxAged data is immediately replaced with fresh data from server.

**Reply by frederikhors** [reply]
Yeah, we have https://github.com/TanStack/query/issues/7049 for this.
